### DISCUSSION TITLE
Cannot install ifm3dpy in container of l4t-tensorflow:r32.4.3-tf2.2-py3

### DISCUSSION THREAD

**Discussion by Petopp** [ifm/ifm3d#402] [Q&A]
Hello all,
thanks to the update for the IFM board, you can now also load larger containers onto it.

But now I have the problem that if I use the container of l4t-tensorflow:r32.4.3-tf2.2-py3 or nvcr.io/nvidia/l4t-base:r32.4.3, I can not install ifm3dpy there.

Does anyone have a tip for me how this could work?

Best regards

**Comment by lola-masson** [comment]
Hi @Petopp, could you provide a bit more details on what goes wrong in the installation? Do you have any error message when you try to pip install?

**Reply by dekhanra** [reply]
@Petopp, I see that the default python interpreter in above mentioned base images is python3.6. To install ifm3dpy via PIP the python version>=3.8 on aarch based systems.

**Reply by lola-masson** [reply]
You will need to build ifm3dpy from source as we do not have pre-built packages for python 3.6.
Follow [the instructions on this page](https://api.ifm3d.com/stable/content/installation_instructions/install_from_source_linux.html#the-default-build), making sure to add the -DBUILD_MODULE_PYBIND11=ON flag to build the python package.

**Reply by Petopp** [reply]
Hello thanks for your answers:

Here are the comletet error message, now with python 3.8


```
 > [4/4] RUN python3.8 -m pip install ifm3dpy:                                                                                                           
#0 3.242 Collecting ifm3dpy
#0 3.657   Could not find a version that satisfies the requirement ifm3dpy (from versions: )
#0 3.662 No matching distribution found for ifm3dpy
------
Dockerfile:9
--------------------
   7 |     RUN apt-get install -y python3.8 python3.8-dev python3-pip
   8 |     
   9 | >>> RUN python3.8 -m pip install ifm3dpy
  10 |     
  11 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c python3.8 -m pip install ifm3dpy" did not complete successfully: exit code: 1
```

**Reply by dekhanra** [reply]
@Petopp, the command you executed is not valid. The better way is to build it from source as lola mentioned. I could give you hint for Dockerfile. This builds the latest stable ifm3d from source. You can also modify this dockerfile to install from debians by extracting zip files mentioned in https://github.com/ifm/ifm3d/releases/tag/v1.2.6 

```dockerfile
FROM nvcr.io/nvidia/l4t-tensorflow:r32.4.3-tf2.2-py3

# Create the ifm user
RUN id ifm 2>/dev/null || useradd --uid 30000 --create-home -s /bin/bash -U ifm
WORKDIR /home/ifm

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y \
    build-essential \
    cmake \
    coreutils \
    findutils \
    git \
    libboost-all-dev \
    libcurl4-openssl-dev \
    libgoogle-glog-dev \
    libgtest-dev \
    libopencv-dev \
    libpcl-dev \
    libproj-dev \
    libxmlrpc-c++8-dev \
    locales \
    ninja-build \
    wget \
    gpg

RUN apt-get clean


ARG CMAKE_VERSION=3.20.6
# Install cmake
RUN wget -O - "https://github.com/Kitware/CMake/releases/download/v${CMAKE_VERSION}/cmake-${CMAKE_VERSION}-linux-$(uname -i).tar.gz" \
    | tar -xz --strip-components=1 -C /usr

# First clone ifm3d repo

ARG IFM3D_CLONE_REPO=https://github.com/ifm/ifm3d
ARG IFM3D_TAG=1.2.6
RUN mkdir src && \
    cd src && \
    git clone ${IFM3D_CLONE_REPO} ifm3d && \
    cd ifm3d && \
    git checkout tags/v${IFM3D_TAG} && \
    echo "Building from tag ${IFM3D_TAG}" && \
    mkdir build && \
    cd build && \
    cmake -GNinja cmake \
    -DCMAKE_INSTALL_PREFIX=/install \
    -DBUILD_MODULE_IMAGE=ON \
    -DBUILD_MODULE_PYBIND11=ON \
    -DBUILD_IN_DEPS=OFF .. && \
    cmake --build . && \
    cmake --build . --target install
```

**Reply by Petopp** [reply]
Thanks for your answer, i will test this way.
